corehq/messaging/tasks.py

```diff
@@ -40,6 +46,16 @@ def sync_case_for_messaging_rule(self, domain, case_id, rule_id):
         self.retry(exc=e)
 
 
+@no_result_task(serializer='pickle', queue=settings.CELERY_REMINDER_CASE_UPDATE_QUEUE, acks_late=True)
+def sync_case_chunk_for_messaging_rule(self, domain, case_id_chunk, rule_id):
+    for case_id in case_id_chunk:
```

Comment:
Batching and processing case ids in chunks seems to imply that you think a significant amount of processing overhead is spent in enqueuing individual `sync_case_for_messaging_rule` tasks. Is that true? Do you have measurements of task processing overhead to back that up?

Unfortunately, since the arguments of `_sync_case_for_messaging_rule` (note leading underscore) are all ids, all of those entities will need to be loaded from the database individually on each iteration of this loop, which makes me think this is not much of an optimization. Have you considered loading the domain object, rule object, and all cases before entering the loop to reduce the number of db hits?

Reply:
> Batching and processing case ids in chunks seems to imply that you think a significant amount of processing overhead is spent in enqueuing individual sync_case_for_messaging_rule tasks. Is that true?

No, that's not my intention and that's not the reason for batching these. Batching is to reduce the amount of tasks triggered to process per number of cases so that we can reduce load on celery. This is necessary because we are queueing at 20x rate with the concurrent shards, and at that rate without batching we would be putting a heavy load on celery where case count is very high like ccs-record.

> Unfortunately, since the arguments of _sync_case_for_messaging_rule (note leading underscore) are all ids, all of those entities will need to be loaded from the database individually on each iteration of this loop, which makes me think this is not much of an optimization. 

Again, this is simply reducing the celery load, everything else being exactly same (such as number of db lookups etc).

Reply:
Daniel said
> a significant amount of processing overhead is spent in enqueuing individual `sync_case_for_messaging_rule` tasks.

Sravan said
> Batching is to reduce the amount of tasks triggered to process per number of cases so that we can reduce load on celery.

I think we said the same thing. This optimization is about reducing celery load. Thanks for clarifying, that makes sense.

Reply:
Alright, hope we are sorted here then.

Reply:
@sravfeyn while I now think I understand your reasons for adding batching, I'm less certain that it will have a measurable impact. Celery is supposed to be able to scale and handle many concurrent tasks. That is its entire job, after all. I'm still curious if you've measured or have any guesses at the task processing overhead you're expecting to save by batching?

To be clear, I am not trying to discourage batching. I think it's a useful first step in allowing more impactful optimizations.

Reply:
The main improvement in this PR is #1 in the PR description, to replace the sequential queueing with concurrent queueing of tasks, this solves the primary bottleneck of iterating through a huge table making it 20 (number of shard times) faster. Batching is only a secondary tweak so that the increased queueing rate doesn't overwhelm celery. Yes, we can scale celery for more high concurrent tasks, but that does come with requiring more resources which I am trying to avoid by batching.

Reply:
We saw about 16X improvement

Reply:
🎖️